Rename UIColor+HexString to UIColor+SDHexString to avoid conflict wit…

ehlersd opened the pull request:
…h other 3rd party libraries

### New Pull Request Checklist

* [X] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [X] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [X] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [X] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [X] I have added the required tests to prove the fix/feature I am adding
* [X] I have updated the documentation (if necessary)
* [X] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: ...

### Pull Request Description

This simply renames the category files for UIColor+HexString to avoid an existing conflict with other 3rd Party cocoapods.

dreampiggy commented:
This is the private API of SDWebImage, and for Objective-C, the **Category Name** does not have any meaning. So what's the main propose of this changes ?

dreampiggy commented:
But if there are something necessary, we can change the header naming. Since it's a private API and can change without any sem-version side-effect.

ehlersd commented:
It’s the filename of the header itself that conflicts with at least one other cocoapod library.  Just like adding the sd_ to the method names and SD to the class names (and class filenames), it should be added to the category filenames to avoid filename collisions.
On Sep 6, 2019, 2:29 AM -0500, DreamPiggy <notifications@github.com>, wrote:
> But if there are something necessary, we can change the header naming. Since it's a private API and can change without any sem-version side-effect.
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

dreampiggy commented:
@ehlerds This header file is `Private Header`. Which means, it does not present in your Header Search Path. So this should not cause anything issue. Because when you write

```
#improt <SDWebImage/UIColor+HexString.h> // This does not compile because of header search
#import "UIColor+HexString.h" // The same
```

dreampiggy commented:
See: https://github.com/SDWebImage/SDWebImage/blob/master/SDWebImage.podspec#L37

Actually, if you have issue, we can change this. But,  you should provide the detailed information or context. For my test, the private header files should not cause any naming conflict or issue. So this is why we just naming it like this.

ehlersd commented:
Actually, the problem was two-fold.  The problem comes in when we have another Cocoapod (in this case iosMath) that has the same category header filename that is not private.  When trying to compile the SDWebImage pod, Xcode loads iosMath's category header instead of the private one included in the SDWebImage pod.   

By making the filename not so generic in SDWebImage, it will reduce the possibility of conflicts like this in the future and ensure that SDWebImage is loading it's actual private header (instead of a different one).  Just like all the rest of the SDWebImage source files are prepended with "SD", so should the category header.

<img width="1527" alt="Screenshot 2019-09-07 11 20 23" src="https://user-images.githubusercontent.com/2302997/64477478-854d9d80-d161-11e9-8bbd-e1ea8a23cd2d.png">

dreampiggy commented:
> Xcode loads iosMath's category header instead of the private one included in the SDWebImage pod.

This only happends, if you custom the header search path generation logic in CocoaPods.

CocoaPods by default, only add the follwing header search path:

```
"${PODS_ROOT}/Headers/Private" "${PODS_ROOT}/Headers/Private/SDWebImage" "${PODS_ROOT}/Headers/Public" "${PODS_ROOT}/Headers/Public/SDWebImage"
```

And, this header directory, does not contains a `UIColor+HexColorString` in the root directory. Which means:

```objectivec
#import "UIColor+HexColorString.h" // This must match the `Pods/Headers/Private/SDWebImage/UIColor+HexColorString, not the `Pods/Headers/Public/iOSMatch/UIColor+HexColorString
``

I can put a demo for you. This is the designed behavior

I think this is the issue that your integration don't use the standard CocoaPods's generation way. For this case, we can not provide and gurantee the offifical help, because they are tons of way for the wrong usage.

We can solve the issue, but my point is that **If you use any custom logic, we do not support by default, you must point out your usage, so that we can find workaround to this**

dreampiggy commented:
[TestPrivateHeaderNamingConflict.zip](https://github.com/SDWebImage/SDWebImage/files/3587258/TestPrivateHeaderNamingConflict.zip)

I uploaded a demo for you. I create one Demo Pod `TestLibrary4`, which contains a file naming `UIColor+HexString.h` and it's the Public Headers. And you can see, we can build both `SDWebImage` and this `TestLibrary4` at the same time.

![image](https://user-images.githubusercontent.com/6919743/64482832-aa0e4780-d22b-11e9-9920-968e21021cd2.png)

ehlersd commented:
This is no custom path that I'm using.   And to demonstrate the problem:
1. Take your zipped example
2. Add back the use_frameworks!
3. run Pod Install
4. try to compile....failure!

<img width="1153" alt="Screenshot 2019-09-07 22 46 51" src="https://user-images.githubusercontent.com/2302997/64483013-6a127a80-d1c1-11e9-8086-8b7a4aeec5c4.png">

ehlersd commented:
And to demonstrate the exact error I'm seeing, I added the iosMath pod:

[TestPrivateHeaderNamingConflictFailure.zip](https://github.com/SDWebImage/SDWebImage/files/3587272/TestPrivateHeaderNamingConflictFailure.zip)

dreampiggy commented:
@ehlersd My fault. The Podfile was wrong because it integrate to the `DemoAppTest` target, not the `DemoApp` target.

I fix it again and run, here is the result:

1. When using `#import "UIColor+HexString.h"`, it can found the header `Pods/Headers/Public/TestLibrary4/UIColor+HexString.h`.

Because there are search path here: 
![image](https://user-images.githubusercontent.com/6919743/64483155-4df9f200-d230-11e9-9233-d0fdd18af561.png)

2. SDWebImage target compile. 

This is what your first screenshot shows build failsure. But during my test, it can compile without any issue.

3. Using the framework `use_frameworks!` will cause a warning:

![image](https://user-images.githubusercontent.com/6919743/64483188-ad580200-d230-11e9-8770-b80317ad6e3e.png)

And in this case (You don't use `@import SDWebImage`), you can search the private header of SDWebImage using `#import <SDWebImage/UIColor+HexString.h>`. This is not the designed behavior and we don't promise the semversion for this import.

ehlersd commented:
If you look at the zip I sent, I corrected your Podfile.  My assumption at this point is that your original test will also fail if you change the name of your pod to one that alphabetically comes BEFORE SDWebImage, therefore would be included first in the Seach Path...

Again, the zip I just sent demonstrates the error using your sample.

dreampiggy commented:
This seems Xcode's bug about the `USE_HEADERMAP` setting. I'm investigating this behavior.

To my knowledge for C/C++ language, this should not happend during the standard clang header search path resolve. Seems Xcode's `use header map` do somthing magic but wrong.

![image](https://user-images.githubusercontent.com/6919743/64484668-40506680-d248-11e9-8e20-badb538e252b.png)

![image](https://user-images.githubusercontent.com/6919743/64484676-77267c80-d248-11e9-90a7-f3d5477db7f2.png)


![image](https://user-images.githubusercontent.com/6919743/64484674-6e35ab00-d248-11e9-8ade-981fa40db64c.png)

dreampiggy commented:
Wait me for one or two days. If I can not find another solution or identify it's a consistent behavior of Xcode and CocoaPods, I can add this to 5.1.2 milestone.

Actually, if we need to change, we can add all the current private category for System Framework Class with the SD prefix. Do them at the same time together.

ehlersd commented:
Gotcha.  This issue has been around for over a year, I've just been manually updating it after each SDWebImage version update.  Just finally got a chance to deal with some PRs to some OpenSource repos (such as this).  

Let me know if I can help more, either with the bughunt, or if you'd like me to help with the renaming.

dreampiggy commented:
@ehlersd I can say this is a `Bug` behavior of Xcode.

When you use CocoaPods to integrate Pods with multiple Xcode Project (Instead of one Xcode Project with Multiple Targets). This strange issue disappears.

Add this line in your Podfile to enable multiple projects from CocoaPods 1.7:

```ruby
install! 'cocoapods', :generate_multiple_pod_projects => true
```

See the compile result:

![image](https://user-images.githubusercontent.com/6919743/64503713-dd6dd680-d2fe-11e9-8253-fbfcc8cb0d08.png)

This time, the header map generated by Xcode, does not contains the wrong other Pod's headers, which are not in the header search path. (There are no `MTLabel.h` in the `SDWebImage-project-headers.hmap`)

![image](https://user-images.githubusercontent.com/6919743/64503749-0ee6a200-d2ff-11e9-94d1-7edb98edb4f4.png)

dreampiggy commented:
The bug can be conluded as:

When building one framework target in a project (which has other framework target as well), the `Use Header Map` build settings will cause Xcode include all the other framework target's public headers. 

When build the same framework target in a project without any framework target (the other framework target is in Xcode's Subproject dependency), the `Use Header Map` build settings does not include all the other framework target's public headers.

dreampiggy commented:
Apple has some documentation about the header map. Maybe this build settings can solve the problem ? (Do not try to flat the headers across all framework targets):

https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW159

ehlersd commented:
Well, updating my podfile seems to fix the issue with the private header, but make Firebase pods unable to build at all.  As I fix one problem after another, it just became too much.  I just reverted everything, switched the podfile back, and I guess I'll just go back to modifying your header file manually after every update.  Sorry to bother you.

bughana commented:
I have the exact same problem... Any chance this PR can be reopened and merged?
Changing the header manually does not really work for my, as I don't directly add SDWebImage as a pod. It is a dependency of another pod I use.

dreampiggy commented:
Solved in [5.7.2](https://github.com/SDWebImage/SDWebImage/releases/tag/5.7.2) version.
